File: frontend/lib/shared/models/social_link.dart
import 'package:equatable/equatable.dart';

/// Plateformes standard supportées.
const List<String> kValidPlatforms = [
  'instagram', 'tiktok', 'youtube', 'linkedin',
  'x', 'facebook', 'strava', 'website',
];

/// Visibilité d'un lien social.
enum LinkVisibility {
  public,
  coachesOnly;

  static LinkVisibility fromString(String s) => s == 'coaches_only'
      ? LinkVisibility.coachesOnly
      : LinkVisibility.public;

  String get value => this == LinkVisibility.coachesOnly ? 'coaches_only' : 'public';
}

/// Lien réseau social d'un utilisateur (coach ou client).
class SocialLink extends Equatable {
  const SocialLink({
    required this.id,
    this.platform,       // null = lien custom
    required this.url,
    this.label,
    this.position = 0,
    this.visibility = LinkVisibility.public,
  });

  final String id;

  /// Slug plateforme (`'instagram'`, `'youtube'`…) ou `null` pour les liens custom.
  final String? platform;

  final String url;

  /// Libellé affiché (obligatoire si platform = null).
  final String? label;

  final int position;
  final LinkVisibility visibility;

  bool get isCustom => platform == null;

  String get displayLabel => label ?? platform ?? 'Lien';

  factory SocialLink.fromJson(Map<String, dynamic> json) => SocialLink(
    id:         json['id'] as String,
    platform:   json['platform'] as String?,
    url:        json['url'] as String,
    label:      json['label'] as String?,
    position:   json['position'] as int? ?? 0,
    visibility: LinkVisibility.fromString(
      json['visibility'] as String? ?? 'public',
    ),
  );

  Map<String, dynamic> toJson() => {
    'id':         id,
    if (platform != null) 'platform': platform,
    'url':        url,
    if (label != null)    'label':    label,
    'position':   position,
    'visibility': visibility.value,
  };

  @override
  List<Object?> get props => [id, platform, url, label, position, visibility];
}

File: backend/app/schemas/social_link.py
"""Schémas Pydantic — Liens réseaux sociaux (Phase 7)."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, field_validator, model_validator

VALID_PLATFORMS = [
    "instagram",
    "tiktok",
    "youtube",
    "linkedin",
    "x",
    "facebook",
    "strava",
    "website",
]

VALID_VISIBILITY = ["public", "coaches_only"]

MAX_LINKS_PER_USER = 20


class SocialLinkCreate(BaseModel):
    """Créer ou remplacer un lien réseau social.

    - platform standard (instagram, tiktok…) → UPSERT, label optionnel
    - platform=None → lien custom, label OBLIGATOIRE
    """

    platform: Optional[str] = None
    label: Optional[str] = None
    url: str
    visibility: str = "public"
    position: int = 0

    @field_validator("platform")
    @classmethod
    def platform_must_be_valid(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_PLATFORMS:
            raise ValueError(
                f"Plateforme invalide. Valeurs acceptées : {', '.join(VALID_PLATFORMS)} ou null pour un lien personnalisé"
            )
        return v

    @field_validator("label")
    @classmethod
    def label_max_length(cls, v: Optional[str]) -> Optional[str]:
        if v and len(v) > 100:
            raise ValueError("Le label ne peut pas dépasser 100 caractères")
        return v

    @field_validator("url")
    @classmethod
    def url_must_be_http(cls, v: str) -> str:
        if not v.startswith("http://") and not v.startswith("https://"):
            raise ValueError("L'URL doit commencer par http:// ou https://")
        if len(v) > 500:
            raise ValueError("L'URL ne peut pas dépasser 500 caractères")
        return v

    @field_validator("visibility")
    @classmethod
    def visibility_must_be_valid(cls, v: str) -> str:
        if v not in VALID_VISIBILITY:
            raise ValueError(
                f"Visibilité invalide. Valeurs acceptées : {', '.join(VALID_VISIBILITY)}"
            )
        return v

    @model_validator(mode="after")
    def label_required_for_custom(self) -> "SocialLinkCreate":
        if self.platform is None and not self.label:
            raise ValueError(
                "Le champ 'label' est obligatoire pour un lien personnalisé (platform=null)"
            )
        return self


class SocialLinkUpdate(BaseModel):
    """Mettre à jour un lien existant (par son ID)."""

    label: Optional[str] = None
    url: Optional[str] = None
    visibility: Optional[str] = None
    position: Optional[int] = None

    @field_validator("url")
    @classmethod
    def url_must_be_http(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            if not v.startswith("http://") and not v.startswith("https://"):
                raise ValueError("L'URL doit commencer par http:// ou https://")
            if len(v) > 500:
                raise ValueError("L'URL ne peut pas dépasser 500 caractères")
        return v

    @field_validator("visibility")
    @classmethod
    def visibility_must_be_valid(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_VISIBILITY:
            raise ValueError(
                f"Visibilité invalide. Valeurs acceptées : {', '.join(VALID_VISIBILITY)}"
            )
        return v

    @field_validator("label")
    @classmethod
    def label_max_length(cls, v: Optional[str]) -> Optional[str]:
        if v and len(v) > 100:
            raise ValueError("Le label ne peut pas dépasser 100 caractères")
        return v


class SocialLinkResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    platform: Optional[str]
    label: Optional[str]
    url: str
    visibility: str
    position: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
